File: src/main/java/com/meghneelgore/geometry/shapes/Polygon.java
/*
 * Copyright (c) 2018.  Meghneel Gore (meghneel.gore@gmail.com)
 */

package com.meghneelgore.geometry.shapes;

/**
 * Interface defining the contract for shapes
 *
 * @author Meghneel Gore meghneel.gore@gmail.com
 */
public interface Polygon {

    /**
     * Returns the perimeter of the shape
     *
     * @return Perimeter in units
     */
    double getPerimeter();

    /**
     * Returns the area of the shape
     *
     * @return Area in square units
     */
    double getArea();

    /**
     * Determines if this shape overlaps another
     *
     * @param polygon The other polygon
     * @return true iff this shape overlaps the other
     */
    boolean overlaps(Polygon polygon);

    /**
     * Returns the {@code PolygonType} of the polygon
     *
     * @return {@code PolygonType.CONVEX} if the polygon is convex, {@code PolygonType.CONCAVE} if the polygon is concave, {@code PolygonType.COMPLEX} otherwise
     */
    PolygonType getPolygonType();

    enum PolygonType {
        CONVEX,
        COCAVE,
        COMPLEX
    }
}

File: src/main/java/com/meghneelgore/geometry/shapes/Polygon.kt
/*
 * Copyright (c) 2018.  Meghneel Gore (meghneel.gore@gmail.com)
 */

package com.meghneelgore.geometry.shapes

import com.google.common.collect.ImmutableList
import com.meghneelgore.geometry.primitives.Segment

/**
 * Interface defining the contract for shapes
 *
 * @author Meghneel Gore meghneel.gore@gmail.com
 */
interface Polygon {

    /**
     * The list of segments making up the polygon
     */
    val segmentsList: ImmutableList<Segment>

    /**
     * Returns the perimeter of the shape
     *
     * @return Perimeter in units
     */
    val perimeter: Double

    /**
     * Returns the area of the shape
     *
     * @return Area in square units
     */
    val area: Double

    /**
     * Returns the `PolygonType` of the polygon
     *
     * @return `PolygonType.Convex` if the polygon is convex, `PolygonType.CONCAVE` if the polygon is concave, `PolygonType.Complex` otherwise
     */
    val polygonType: PolygonType

    /**
     * Determines if this shape overlaps another
     *
     * @param polygon The other polygon
     * @return true iff this shape overlaps the other
     */
    fun overlaps(polygon: Polygon): Boolean

    enum class PolygonType {
        Convex,
        Concave,
        Complex
    }
}
